mobile/android/android-emulator-services.ts

```diff
@@ -120,29 +147,113 @@ class AndroidEmulatorServices implements Mobile.IEmulatorPlatformServices {
+			return model;
+		}).future<string>()();
+	}
+
+	private getNameFromNativeEmulatorId(emulatorId: string): IFuture<string> {
+		var match = emulatorId.match(AndroidEmulatorServices.RUNNING_ANDROID_EMULATOR_REGEX);
+		var portNumber: string;
+		if(match && match[1]) {
+			portNumber = match[1];
+		} else {
+			Future.fromResult("");
+		}
+
+		var future = new Future<string>();
+		var output: string = "";
+		var client = net.connect({ port: portNumber }, () => { 
+			client.write(util.format("avd name%s", os.EOL));
+		});
+
+		client.on('data', (data: any) => {
+			output += data.toString();
+			client.end();
+		});
+		client.on('end', () => {
+			//output should look like:
+			//Android Console: type 'help' for a list of commands
+			//OK
+			//<Name of image>
+			//OK
+			//
+			var name: string;
+			var foundOK = false;
+			var lines: string[] = output.split(os.EOL);
+			// find line between OK
+			_(lines).each((line: string) => {
+				if(foundOK) {
+					name = line.trim();
+					return false;
+				} else if(line.match(/^OK/)) {
+					foundOK = true;
+				}
+			});
+
+			future.return(name);
+		});
+		return future;
+	}
+
 	private startEmulatorInstance(image: string): IFuture<string> {
 		return (() => {
-			var initiallyRunningEmulators = this.getRunningEmulators().wait();
+			var emulatorId = this.getRunningEmulatorId(image).wait();
+			this.endTimeEpoch = helpers.getCurrentEpochTime() + this.getMilliSecondsTimeout();
+			if(emulatorId) {
+				// If there's already a running instance of this image, we'll jusr deploy the app to it.
+				return emulatorId;
+			}
 
-			// If there's no running emulators or the user had specified an image (--avd) we have to start new emulator.
-			if(initiallyRunningEmulators.length === 0 || options.avd) {
-				this.$logger.info("Starting Android emulator with image %s", image);
+			// have to start new emulator
+			this.$logger.info("Starting Android emulator with image %s", image);
+			if(options.geny) {
+				//player is part of Genymotion, it should be part of the PATH.
+				this.$childProcess.spawn("player", ["--vm-name", image],
+					{ stdio: "ignore", detached: true }).unref();
+			} else {
 				this.$childProcess.spawn('emulator', ['-avd', image],
 					{ stdio: "ignore", detached: true }).unref();
 			}
-
-			var runningEmulators = this.getRunningEmulators().wait();
+			
 			var isInfiniteWait = this.getMilliSecondsTimeout() === 0;
-			this.endTimeEpoch = helpers.getCurrentEpochTime() + this.getMilliSecondsTimeout();
 			var hasTimeLeft = helpers.getCurrentEpochTime() < this.endTimeEpoch;
 
 			while(hasTimeLeft || isInfiniteWait) {
-				if(runningEmulators.length > initiallyRunningEmulators.length || (runningEmulators.length > 0 && !options.avd)) {
-					break;
+				emulatorId = this.getRunningEmulatorId(image).wait();
+				if(emulatorId) {
+					return emulatorId;
 				}
 
 				this.sleep(10000); // the emulator definitely takes its time to wake up
```

Comment:
I would postpone until it is needed anywhere else - the utilities lib is already a bag of all tricks